packages/e2e-test-utils/src/rest-api.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * External dependencies
+ */
+import fetch from 'node-fetch';
+
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+
+/**
+ * Internal dependencies
+ */
+import { WP_BASE_URL, WP_USERNAME, WP_PASSWORD } from './shared/config';
+
+// `apiFetch` expects `window.fetch` to be available in its default handler.
+global.window = global.window || {};
+global.window.fetch = fetch;
+
+const setAPIRootURL = ( async () => {
+	// Discover the API root url using link header.
+	// See https://developer.wordpress.org/rest-api/using-the-rest-api/discovery/#link-header
+	const res = await fetch( WP_BASE_URL, { method: 'HEAD' } );
+	const links = res.headers.get( 'link' );
+	const restLink = links.match( /<([^>]+)>; rel="https:\/\/api\.w\.org\/"/ );
+
+	if ( ! restLink ) {
+		throw new Error( `Failed to discover REST API endpoint.
+Link header: ${ links }` );
+	}
+
+	const rootURL = restLink[ 1 ];
+	apiFetch.use( apiFetch.createRootURLMiddleware( rootURL ) );
+} )();
+
+/**
+ * Call REST API using `apiFetch` to build and clear test states.
+ *
+ * @param {Object} options `apiFetch` options.
+ * @return {Promise<any>} The response value.
+ */
+async function rest( options = {} ) {
+	await setAPIRootURL;
+
+	return await apiFetch( {
+		...options,
+		headers: {
+			...( options.headers || {} ),
+			// Authorize via basic authentication and test plugin.
+			Authorization: `Basic ${ Buffer.from(
```

Comment:
There's a caveat here that these utilities only work with the basic auth plugin, so it might be good to make that clear in the function docs for `rest` and `batch` (for when they're potentially non-experimental).

That dependency is an interesting one. A thought that occurs to me is how the rest utils might work in another test environment, e.g. WordPress core. I suppose the `e2e-tests` package could be imported, but feels like the plugin or its code should almost be part of the test-utils package.

Reply:
Yeah, good point! I refactored it to use cookie/nonce to authenticate the API requests, now we don't need the plugin for `rest` and `batch` to work.